### PR TITLE
JVM_IR: generate function reference classes before property references

### PR DESCRIPTION
**Pull request by pyos**
This ensures that the `$$delegatedProperties` array is placed into the class in which it is used. It also removes the need for the lambda-handling part of AddContinuationLowering to support visibilities and bound parameters.

 #KT-42253 Fixed
 #KT-39434 Fixed

### CODE REVIEW

**Comment by udalov** [comment]
Looks like this also fixes [KT-40464](https://youtrack.jetbrains.com/issue/KT-40464). Could you please unmute `IrKotlinSteppingTestGenerated$Custom.testCoroutine`?

**Comment by pyos** [comment]
Done. I'll take your word that the test should be unmuted because I can't run the debugger tests right now for some reason (`java.lang.InternalError: java.io.IOException: usr/local/buildtools/java/jdk11 not found`).

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt [hunk 1]
```diff
@@ -289,6 +290,8 @@ private val jvmFilePhases = listOf(
 
     moveOrCopyCompanionObjectFieldsPhase,
     inlineCallableReferenceToLambdaPhase,
+    functionReferencePhase,
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:293]: Maybe add `functionReferencePhase` as a `prerequisite` for `propertyReferencePhase`?

Comment (pyos) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:293]: Done, `suspendLambdaPhase` should be there too I guess.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AddContinuationLowering.kt [hunk 2]
```diff
@@ -36,201 +36,172 @@ import org.jetbrains.kotlin.ir.expressions.impl.*
...
+        override fun visitGetValue(expression: IrGetValue) {
+            result = result || (expression.isReadOfCrossinline() && expression.symbol.owner.parent !in functions)
+        }
+
+        private val functions = mutableSetOf<IrFunction>()
+    }, null)
+    return result
+}
+
+private class SuspendLambdaLowering(context: JvmBackendContext) : SuspendLoweringBase(context) {
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AddContinuationLowering.kt:115]: IMHO moving it to a separate file would be more readable (and also maybe `SuspendLoweringBase` too)

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrInlineUtils.kt [hunk 3]
```diff
@@ -69,8 +69,8 @@ private fun IrBody.move(
     override fun visitBlock(expression: IrBlock): IrExpression {
         // Might be an inline lambda argument; if the function has already been moved out, visit it explicitly.
         if (expression.origin == IrStatementOrigin.LAMBDA || expression.origin == IrStatementOrigin.ANONYMOUS_FUNCTION)
-            if (expression.statements[0] !is IrFunction && expression.statements[1] is IrFunctionReference)
-                (expression.statements[1] as IrFunctionReference).symbol.owner.transformChildrenVoid()
+            if (expression.statements.lastOrNull() is IrFunctionReference && !expression.statements.any { it is IrFunction })
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrInlineUtils.kt:72]: `!any` -> `none`

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AddContinuationLowering.kt (hunk 2 above)
Comment (pyos) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AddContinuationLowering.kt]: Done. `SuspendLoweringBase` is too short for that I think, it should look fine with `SuspendLambdaLowering`.

**Review by ilmirus — APPROVED** [review]

**Comment by udalov** [comment]
Thank you!

**Comment by udalov** [comment]
@pyos Found a regression: https://youtrack.jetbrains.com/issue/KT-42527

**Comment by udalov** [comment]
@pyos Another regression: https://youtrack.jetbrains.com/issue/KT-42554

**Comment by udalov** [comment]
Another one: https://youtrack.jetbrains.com/issue/KT-42562

**Pull request merged**